### ISSUE TITLE
No document formatter for 'django-html'-files installed.

### ISSUE THREAD

**Issue by FSE-DEV** [vscode-django/vscode-django#23]
With extension enabled and Language Mode set to "Django HTML" (auto) formatting the document or selections of it stopped working.
VSC Info Message: There is no document formatter for 'django-html'-files installed.

I guess the formatter could still follow the same set of rules with django-html as it could with regular html. Only, I am not aware of any setting that would allow to achieve such connection.

**Comment by rtpm** [comment]
I am not sure if default HTML vscode formatter is able to format django html template properly.
Problably js-beautify ?

I've got lots of problems with templatetags splitted into new lines or empty spaces put after quotes. Example:

`<a href="{%url "someView" var1=someVar %}">link<?a>`

If someone knows the solution I would really be grateful.
Thanks

**Comment by thibault** [comment]
As a reference purpose, a solution can be found [on this thread](https://github.com/Microsoft/vscode/issues/65189).

**Comment by msurdi** [comment]
@thibault That discussion seems to be about associating html files with the django-html file type. Even after doing that, there is still no formatter for django templates. I've been looking around and it seems that js-beautify doesn't support Django's template syntax.

**Comment by thibault** [comment]
@msurdi You are right. However, I've found that vscode's default html formatter results are not too bad on django template files, so to me it's an acceptable temporary workaround, and that's why I linked that discussion.

I agree that a better formatter would be neat, but since it's open source and I'm not willing to do it myself, well, I'm just rolling with what I have.

**Comment by msurdi** [comment]
I'm having troubles with the default formatter, like inserting spaces inside quoted strings, etc. But after looking further into this, I think the problem might have to be related to the js-beautify version. It looks like the beautify extension is using 1.8.8 (I'm looking [here](https://github.com/HookyQR/VSCodeBeautify/blob/739bc9a165e2522d0c87b6169d516614f522c1c1/package.json#L209).

When trying from the command line formatting the Django templates like `js-beautify -r *.html` with a manually installed (with `npm i -g js-beautify@next`) it seems to be doing a much better job. Also, https://beautifier.io/ seems to work quite well.

Just leaving this here for future people searching for this.

**Comment by IEB-Richard** [comment]
Here is my solution:

first, install HookyQR Beautify extension, and make the following configurations to your VSCode user settings:

```
    "files.associations": {
        "**/templates/*.html": "django-html",
        "**/templates/*": "django-txt",
        "**/requirements{/**,*}.{txt,in}": "pip-requirements"
    },
```
this is associated your django-txt and django-html file as a html file. 

Then add the following two your user settings:

```
    "beautify.language": {
        "html": [
            "htm", "html", "django-html"
        ]
    },
```

**Comment by HymanZHAN** [comment]
Adding to @IEB-Richard 's point, you can also apply `django-html`-specific configuration:
```
{
// your other configs

    "[django-html]": {
        "editor.formatOnSave": true,
        "editor.defaultFormatter": "HookyQR.beautify"
    }
}
``` 
So that you can format the template on save.

**Comment by ZurMaD** [comment]
It works on settings.json added.
```

{
    "python.linting.pylintArgs": [
        "--load-plugins=pylint_django"
    ],
    "files.associations": {
        "**/templates/*.html": "django-html",
        "**/templates/*": "django-txt",
        "**/requirements{/**,*}.{txt,in}": "pip-requirements"
    },
    "beautify.language": {
        "html": [
            "htm", "html", "django-html"
        ]
    },
}
```

**Comment by IEB-Richard** [comment]
@ZurMaD 

Thank you, very helpful settings!

**Comment by parmar-amit** [comment]
Thank you, this is really helpful. Much appreciated!

**Comment by IEB-Richard** [comment]
I've already forgotten this thread and the settings I made in last Dec. 

Today, when I setup a new Macbook Pro,I meet the same issue again. I search on the internet, and find this thread again. LOL! 

Thank you, everybody!

**Comment by david-daming** [comment]
Thank you guys, this is really helpful. I like this thread! 

From after the reply of @IEB-Richard 
everyone put some inputs, and collect all your inputs, I finally got a perfect solution!

To make this thread more helpful, I collect all-in-one here:

Step-01: install the extension Beautify from HookyQR
you can easily search it from the vscode extension. 

Step-02: under your project folder, find .vscode/settings.json and make the following configurations:

```
{

  // other settings
  "files.associations": {
    "**/*.html": "html",
    "**/templates/**/*.html": "django-html",
    "**/templates/**/*": "django-txt",
    "**/requirements{/**,*}.{txt,in}": "pip-requirements"
  },
  "emmet.includeLanguages": {
    "django-html": "html"
  },
  "beautify.language": {
    "html": [
      "htm", "html", "django-html"
    ]
  },
  "[django-html]": {
    "editor.formatOnSave": true,
    "editor.defaultFormatter": "HookyQR.beautify"
  }
}
```

**Comment by sankalpmukim** [comment]
prettier/prettier#5581 has a solution to all the above problems. It's a perfect solution.

**Comment by ilahazs** [comment]
> [prettier/prettier#5581](https://github.com/prettier/prettier/issues/5581) has a solution to all the above problems. It's a perfect solution.

Prettier is kinda bad if u used it on django-html files. 

Look this:
Beautify:
![image](https://github.com/vscode-django/vscode-django/assets/68785569/e76a2cef-363c-4a3f-a2c4-d014becaab82)

Prettier:
![image](https://github.com/vscode-django/vscode-django/assets/68785569/b35fd65f-397b-4fe1-8dec-7bba5d810a62)

**Comment by IArnaut2** [comment]
> Here is my solution:
> 
> first, install HookyQR Beautify extension, and make the following configurations to your VSCode user settings:
> 
> ```
>     "files.associations": {
>         "**/templates/*.html": "django-html",
>         "**/templates/*": "django-txt",
>         "**/requirements{/**,*}.{txt,in}": "pip-requirements"
>     },
> ```
> 
> this is to associatedyour django-txt and django-html file as a html file.
> 
> Then add the following two your user settings:
> 
> ```
>     "beautify.language": {
>         "html": [
>             "htm", "html", "django-html"
>         ]
>     },
> ```
> 
> After that, you can format your django-html file with command alt+shift+f to format your code.

That extension is deprecated.
